apache/metron
# METRON-1685: Retrieve Pcap results in raw binary format

## Contributor Comments
This PR adds a REST endpoint that allows a user to download a pcap page in HDFS as a file to their local file system. This will allow users to save files for further investigation outside of Metron, in a tool like wireshark.

A new getRawPcap method was added to the PcapService. This function gets a HDFS input stream given a path that's looked up by username, job id and page number. The stream is then copied to the http response outputstream and the appropriate response headers are set.  Filenames follow the format `pcap_{jobId}_{pageNumber}.pcap` but can be overriden by the `fileName` request parameter.

### Changes Included
- A new Pcap download endpoint that accepts a job id and page number and returns a file for download
- A new PcapService getRawPcap method that returns an input stream from an HDFS path given a username, job id and page number
- Unit and integration tests

### Testing
This has been tested in full dev. Testing instructions are as follows:

1. Spin up full dev
2. Create the HDFS directories specified in application.yml
  - /apps/metron/pcap/input
  - /apps/metron/pcap/interim
  - /apps/metron/pcap/output
3. Put pcap data in `/apps/metron/pcap/input`
4. Run a fixed pcap query to generate results:
```
curl -X POST --header 'Content-Type: application/json' --header 'Accept: application/json' -d '{
  "endTimeMs": 1458240269424,
  "startTimeMs": 1458240269419
}' 'http://node1:8082/api/v1/pcap/fixed'
```
5. Monitor the job with the get status endpoint until the job has SUCCEEDED and returns a pageTotal:
```
{
  "jobId": "job_1532012905403_0001",
  "jobStatus": "SUCCEEDED",
  "description": "map: 20.0%, reduce: 0.0%",
  "percentComplete": 100,
  "pageTotal": 2
}
```
6. Download a file for a job and page number:
```
curl -X GET --header 'Accept: application/octet-stream' 'http://node1:8082/api/v1/pcap/job_1532012905403_0001/raw?page=1'
```
This should download a file named `pcap_job_1532012905403_0001_1.pcap' to your local file system (job id will match yours).

7. Download a file for a job and page number with a custom file name `pcap1.pcap`:
```
curl -X GET --header 'Accept: application/octet-stream' 'http://node1:8082/api/v1/pcap/job_1532012905403_0001/raw?page=1&fileName=pcap1.pcap'
```
This should download a file named `pcap1.pcap`.

8. Attempt to download a file for a job or page number that doesn't exist.  You should get a 404.

I also verified that the file sizes of the files downloaded and the results in HDFS match.

## Pull Request Checklist

Thank you for submitting a contribution to Apache Metron.  
Please refer to our [Development Guidelines](https://cwiki.apache.org/confluence/pages/viewpage.action?pageId=61332235) for the complete guide to follow for contributions.  
Please refer also to our [Build Verification Guidelines](https://cwiki.apache.org/confluence/display/METRON/Verifying+Builds?show-miniview) for complete smoke testing guides.  


In order to streamline the review of the contribution we ask you follow these guidelines and ask you to double check the following:

### For all changes:


### For code changes:
  ```
  mvn -q clean integration-test install && dev-utilities/build-utils/verify_licenses.sh 
  ```


### For documentation related changes:

  ```
  cd site-book
  mvn site
  ```

#### Note:
Please ensure that once the PR is submitted, you check travis-ci for build issues and submit an update to your PR as soon as possible.
It is also recommended that [travis-ci](https://travis-ci.org) is set up for your personal repository such that your branches are built there before submitting a pull request.
